experiment/oodatamodel/oodatamodel.go

```diff
@@ -0,0 +1,156 @@
+type HTTPTor struct {
+	ExitIP   *string `json:"exit_ip"`
+	ExitName *string `json:"exit_name"`
+	IsTor    bool    `json:"is_tor"`
+}
+
+// HTTPBody is an HTTP body. We use this helper class to define a custom
+// JSON encoder that allows us to choose the proper representation depending
+// on whether the Value field is UTF-8 or not.
+type HTTPBody struct {
+	Value string
+}
+
+// MarshalJSON marshal the body to JSON following the OONI spec that says
+// that UTF-8 bodies are represened as string and non-UTF-8 bodies are
+// instead represented as `{"format":"base64","data":"..."}`.
+func (hb HTTPBody) MarshalJSON() ([]byte, error) {
+	if utf8.ValidString(hb.Value) {
+		return json.Marshal(hb.Value)
+	}
+	er := make(map[string]string)
+	er["format"] = "base64"
+	er["data"] = base64.StdEncoding.EncodeToString([]byte(hb.Value))
+	return json.Marshal(er)
+}
+
+// HTTPRequest contains an HTTP request
+type HTTPRequest struct {
+	Body        HTTPBody          `json:"body"`
+	Headers     map[string]string `json:"headers"`
+	Method      string            `json:"method"`
+	Tor         HTTPTor           `json:"tor"`
+	URL         string            `json:"url"`
+	XBodyIsSnap bool              `json:"x_body_is_snap"`
+}
+
+// HTTPResponse contains an HTTP response
+type HTTPResponse struct {
+	Body        HTTPBody          `json:"body"`
+	Code        int64             `json:"code"`
+	Headers     map[string]string `json:"headers"`
+	XBodyIsSnap bool              `json:"x_body_is_snap"`
+}
+
+// RequestEntry is one of the entries that are part of
+// the "requests" key of a OONI report.
+type RequestEntry struct {
+	Failure  *string      `json:"failure"`
+	Request  HTTPRequest  `json:"request"`
+	Response HTTPResponse `json:"response"`
+}
+
+// RequestList is a list of RequestEntry
+type RequestList []RequestEntry
+
+// NewRequestList returns the list for "requests"
+func NewRequestList(httpresults *porcelain.HTTPDoResults) RequestList {
+	// TODO(bassosimone): here I'm using netx snapshots which are
+	// limited to 1<<20. They are probably good enough for a really
+	// wide range of cases, and truncating the body seems good for
+	// loading measurements on mobile as well. I think I should make
+	// sure I modify the documentation to mention that.
+	var out RequestList
+	if httpresults == nil {
+		return out
+	}
+	in := httpresults.TestKeys.HTTPRequests
+	// OONI's data format wants more recent request first
+	for idx := len(in) - 1; idx >= 0; idx-- {
+		var entry RequestEntry
+		entry.Failure = makeFailure(in[idx].Error)
+		entry.Request.Headers = make(map[string]string)
+		for key, values := range in[idx].RequestHeaders {
+			for _, value := range values {
+				entry.Request.Headers[key] = value
+				break // known data format issue
+			}
+		}
+		entry.Request.Method = in[idx].RequestMethod
+		entry.Request.URL = in[idx].RequestURL
+		entry.Request.Body.Value = string(in[idx].RequestBodySnap)
+		entry.Request.XBodyIsSnap = in[idx].MaxBodySnapSize > 0 &&
```

Comment:
What is this?

Reply:
The code in netx does not record the whole body but a snapshot. The size of the snapshot is configurable. The default size is `1<<20`. All bodies larger than that (or than the configured snapshot size) are snapshots, and this is indicated in the data model. The related pull request is the following: https://github.com/ooni/netx/pull/123.

Reply:
Ah ok I see. Then I think this variable naming is pretty poor. 
Snap means so many different things. The first thing I thought of was that it had something to do with snappy compression, but then also perhaps the snap store for snappy packages. Maybe you can fully expand it and say `Snapshot` or even better maybe `Truncated`?

Variable names aside, are we sure we want the default to be like this for all HTTP requests? This will have an impact on the data processing side of things and would like input from @FedericoCeratto on this too.

Reply:
Sure, I can rename this variable as `BodyIsTruncated`. I have moved the discussion regarding whether or not to truncate the body into the `#ooni-dev` channel, because it's probably beneficial to have a more interactive discussion on this topic.

Reply:
Update: I've discussed this with @FedericoCeratto on Slack and he said the following:

1. truncating the body may help to optimize uploading and processing

2. he doubts fingerprints are after `1<<20` within the body but we should probably check

Reply:
Variable renamed as `BodyIsTruncated`. Documented here: https://github.com/ooni/spec/commit/fcabb0a9b42a83c3360401fda9bd4bb44cc51cdb